contracts/src/v0.8/dev/keeper2_0/KeeperRegistry2_0.sol

```diff
@@ -40,30 +44,691 @@ contract KeeperRegistry2_0 is
+    return _performUpkeep(upkeep, performData);
+  }
+
+  /**
+   * @notice uses LINK's transferAndCall to LINK and add funding to an upkeep
+   * @dev safe to cast uint256 to uint96 as total LINK supply is under UINT96MAX
+   * @param sender the account which transferred the funds
+   * @param amount number of LINK transfer
+   */
+  function onTokenTransfer(
+    address sender,
+    uint256 amount,
+    bytes calldata data
+  ) external override {
+    if (msg.sender != address(i_link)) revert OnlyCallableByLINKToken();
+    if (data.length != 32) revert InvalidDataLength();
+    uint256 id = abi.decode(data, (uint256));
+    if (s_upkeep[id].maxValidBlocknumber != UINT32_MAX) revert UpkeepCancelled();
+
+    s_upkeep[id].balance = s_upkeep[id].balance + uint96(amount);
+    s_storage.expectedLinkBalance = s_storage.expectedLinkBalance + amount;
+
+    emit FundsAdded(id, sender, uint96(amount));
+  }
+
+  ////////
+  // SETTERS
+  ////////
+
+  /**
+   * @inheritdoc OCR2Abstract
+   */
+  function setConfig(
+    address[] memory signers,
+    address[] memory transmitters,
+    uint8 f,
+    bytes memory onchainConfig,
+    uint64 offchainConfigVersion,
+    bytes memory offchainConfig
+  ) external override onlyOwner {
+    if (signers.length > maxNumOracles) revert TooManyOracles();
+    if (f == 0) revert IncorrectNumberOfFaultyOracles();
+    if (signers.length != transmitters.length || signers.length <= 3 * f) revert IncorrectNumberOfSigners();
+
+    // remove any old signer/transmitter addresses
+    uint256 oldLength = s_signersList.length;
+    address signer;
+    address transmitter;
+    for (uint256 i = 0; i < oldLength; i++) {
+      signer = s_signersList[i];
+      transmitter = s_transmittersList[i];
+      delete s_signers[signer];
+      // Do not delete the whole transmitter struct as it has balance information stored
+      s_transmitters[transmitter].active = false;
+    }
+    delete s_signersList;
+    delete s_transmittersList;
+
+    // add new signer/transmitter addresses
+    for (uint256 i = 0; i < signers.length; i++) {
+      if (s_signers[signers[i]].active) revert RepeatedSigner();
+      s_signers[signers[i]] = Signer({active: true, index: uint8(i)});
+
+      if (s_transmitters[transmitters[i]].active) revert RepeatedTransmitter();
+      s_transmitters[transmitters[i]].active = true;
+      s_transmitters[transmitters[i]].index = uint8(i);
+    }
+    s_signersList = signers;
+    s_transmittersList = transmitters;
+
+    // Set the onchain config
+    OnchainConfig memory onchainConfigStruct = abi.decode(onchainConfig, (OnchainConfig));
+    if (onchainConfigStruct.maxPerformGas < s_storage.maxPerformGas) revert GasLimitCanOnlyIncrease();
+    if (onchainConfigStruct.maxCheckDataSize < s_storage.maxCheckDataSize) revert MaxCheckDataSizeCanOnlyIncrease();
+    if (onchainConfigStruct.maxPerformDataSize < s_storage.maxPerformDataSize)
+      revert MaxPerformDataSizeCanOnlyIncrease();
```

Comment:
I added these two for consistency with gasLimitCanIncrease but have mixed feelings about it. Can remove all of them to simplify